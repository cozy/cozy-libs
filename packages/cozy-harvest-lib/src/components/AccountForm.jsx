import React, { PureComponent } from 'react'
import { Form, Field as FinalFormField } from 'react-final-form'
import PropTypes from 'react-proptypes'

import Button from 'cozy-ui/react/Button'
import { translate, extend } from 'cozy-ui/react/I18n'
import Field from 'cozy-ui/react/Field'

import AccountFormError from './AccountFormError'
import {
  getEncryptedFieldName,
  getFieldPlaceholder,
  sanitizeSelectProps
} from '../helpers/fields'
import { KonnectorJobError } from '../helpers/konnectors'
import manifest from 'helpers/manifest'
import OAuthForm from './OAuthForm'

const predefinedLabels = [
  'answer',
  'birthdate',
  'code',
  'date',
  'email',
  'firstname',
  'lastname',
  'login',
  'password',
  'phone'
]

// Out of scope labels already used, should be transferred directly in manifests
// in the future.
const legacyLabels = [
  'branchName' // Used in banking konnectors
]

const VALIDATION_ERROR_REQUIRED_FIELD = 'VALIDATION_ERROR_REQUIRED_FIELD'

export class AccountField extends PureComponent {
  render() {
    const {
      disabled,
      hasError,
      initialValue,
      label,
      name,
      onKeyUp,
      required,
      role,
      t,
      type
    } = this.props

    // Allow manifest to specify predefined label
    const localeKey =
      predefinedLabels.includes(label) || legacyLabels.includes(label)
        ? label
        : name

    const isEditable = !(role === 'identifier' && initialValue)

    const fieldProps = {
      ...this.props,
      autoComplete: 'off',
      className: 'u-m-0', // 0 margin
      disabled: disabled || !isEditable,
      error: hasError,
      fullwidth: true,
      label: t(`fields.${localeKey}.label`, {
        _: t(`legacy.fields.${localeKey}.label`, { _: name })
      }),
      onKeyUp: onKeyUp,
      placeholder: getFieldPlaceholder(
        this.props,
        t(`fields.${name}.placeholder`, { _: '' })
      ),
      side: !required && t('accountForm.fields.optional'),
      size: 'medium'
    }
    const passwordLabels = {
      hideLabel: t('accountForm.password.hide'),
      showLabel: t('accountForm.password.show')
    }
    switch (type) {
      case 'date':
        return (
          <Field
            {...fieldProps}
            placeholder={t(`fields.${name}.placeholder`, {
              _: t(`default.dateFormat`).toLowerCase()
            })}
          />
        )
      case 'dropdown':
        return <Field {...sanitizeSelectProps(fieldProps)} />
      case 'password':
        return (
          <Field
            {...fieldProps}
            // Usting the `new-password` value is the best way to avoid
            // autocomplete for password.
            // See https://stackoverflow.com/a/17721462/1135990
            autoComplete="new-password"
            secondaryLabels={passwordLabels}
          />
        )
      default:
        return <Field {...fieldProps} type="text" />
    }
  }
}

AccountField.propTypes = {
  hasError: PropTypes.bool,
  initialValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  onKeyUp: PropTypes.func,
  role: PropTypes.string,
  type: PropTypes.oneOf(['date', 'dropdown', 'email', 'password', 'text']),
  t: PropTypes.func
}

// As SelectBox component from Cozy-UI, rendering dropdown type, is just giving
// us the full Option object, we just get its value to facilitate mapping
// with account
const parse = type => value => {
  return type === 'dropdown' ? value.value : value
}

export class AccountFields extends PureComponent {
  render() {
    const {
      container,
      disabled,
      hasError,
      initialValues,
      manifestFields,
      onKeyUp,
      t
    } = this.props

    // Ready to use named fields array
    const namedFields = Object.keys(manifestFields).map(fieldName => ({
      ...manifestFields[fieldName],
      name: fieldName
    }))

    return (
      <div>
        {namedFields.map((field, index) => (
          <FinalFormField
            key={index}
            name={field.name}
            parse={parse(field.type)}
          >
            {({ input }) => (
              <AccountField
                {...field}
                {...input}
                container={container}
                hasError={hasError}
                disabled={disabled}
                initialValue={
                  initialValues[field.name] ||
                  initialValues[getEncryptedFieldName(field.name)]
                }
                onKeyUp={onKeyUp}
                t={t}
              />
            )}
          </FinalFormField>
        ))}
      </div>
    )
  }
}

AccountFields.propTypes = {
  disabled: PropTypes.bool,
  hasError: PropTypes.bool,
  fillEncrypted: PropTypes.bool,
  manifestFields: PropTypes.object.isRequired,
  onKeyUp: PropTypes.func,
  t: PropTypes.func
}

export class AccountForm extends PureComponent {
  constructor(props, context) {
    super(props, context)
    const { konnector, lang } = props
    const { locales } = konnector
    if (locales && lang) {
      extend(locales[lang])
    }
  }

  /**
   * Indicates if the state of ReactFinalForm implies that data can be submitted
   * @param  {Object} formState See https://github.com/final-form/final-form#formstate
   * @return {Boolean}
   */
  isSubmittable({ dirty, error, initialValues, valid }) {
    const untouched = initialValues && !dirty
    return error || (valid && !untouched)
  }

  handleKeyUp(event, { dirty, form, initialValues, valid, values }) {
    if (
      event.code === 'Enter' &&
      this.isSubmittable({ dirty, initialValues, valid })
    ) {
      this.handleSubmit(values, form)
    }
  }

  handleSubmit(values, form) {
    const { onSubmit } = this.props
    // Reset form with new values to set back dirty to false
    form.reset(values)
    onSubmit(values)
  }

  validate = (fields, initialValues) => vals => {
    let errors = {}
    for (let name in fields)
      if (
        fields[name].required &&
        !vals[name] &&
        // Don't require value for empty encrypted fields with initial value
        !initialValues[getEncryptedFieldName(name)]
      )
        errors[name] = VALIDATION_ERROR_REQUIRED_FIELD
    return errors
  }

  render() {
    const { account, error, konnector, onSubmit, submitting, t } = this.props
    const { fields, oauth } = konnector

    if (oauth) return <OAuthForm initialValues={initialValues} oauth={oauth} />

    const sanitizedFields = manifest.sanitizeFields(fields)
    const defaultValues = manifest.defaultFieldsValues(sanitizedFields)
    const initialValues = account && account.auth
    const initialAndDefaultValues = { ...defaultValues, ...initialValues }

    let container = null

    return (
      <Form
        initialValues={initialAndDefaultValues}
        onSubmit={onSubmit}
        validate={this.validate(sanitizedFields, initialAndDefaultValues)}
        render={({ dirty, form, values, valid }) => (
          <div
            ref={element => {
              container = element
            }}
          >
            {error && (
              <AccountFormError error={error} konnector={konnector} t={t} />
            )}
            <AccountFields
              container={container}
              disabled={submitting}
              hasError={
                error &&
                error instanceof KonnectorJobError &&
                error.isLoginError()
              }
              initialValues={initialAndDefaultValues}
              manifestFields={sanitizedFields}
              onKeyUp={event =>
                this.handleKeyUp(event, {
                  dirty,
                  form,
                  initialValues,
                  valid,
                  values
                })
              }
              t={t}
            />
            <Button
              busy={submitting}
              className="u-mt-2 u-mb-1-half"
              disabled={
                submitting ||
                !this.isSubmittable({ dirty, error, initialValues, valid })
              }
              extension="full"
              label={t('accountForm.submit.label')}
              onClick={() => this.handleSubmit(values, form)}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  account: PropTypes.object,
  konnector: PropTypes.object.isRequired,
  error: PropTypes.object,
  submitting: PropTypes.bool
}

export default translate()(AccountForm)
