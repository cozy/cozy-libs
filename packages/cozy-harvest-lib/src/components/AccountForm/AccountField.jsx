import Field from 'cozy-ui/transpiled/react/Field'
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'

import { translate } from 'cozy-ui/transpiled/react/I18n'

import { getFieldPlaceholder, sanitizeSelectProps } from '../../helpers/fields'
import {
  legacyLabels,
  predefinedLabels,
  ROLE_IDENTIFIER
} from '../../helpers/manifest'
import { isIOS } from 'cozy-device-helper'

// On iOS, focusing an input not in the same tick as the user action is buggy.
// The field is focused but the keyboard does not show up,
// See https://github.com/apache/cordova-plugin-wkwebview-engine/pull/37/#issuecomment-308321094
const canAutofocus = () => {
  if (isIOS()) {
    return false
  }
  return true
}

/**
 * AccountField encapsulate an unique Cozy-UI Field component.
 * It maps its own props to Cozy-UI Field ones, and deal with i18n and locale
 * logic.
 *
 * AccountField also expect to receive all properties defined in manifest
 * fields.
 */
export class AccountField extends PureComponent {
  constructor(props) {
    super(props)
    // Ref to identifier input, needed to give focus to identifier
    this.inputRef = null
    this.setInputRef = this.setInputRef.bind(this)
  }

  componentDidMount() {
    const { role } = this.props
    if (role === ROLE_IDENTIFIER && this.inputRef && canAutofocus()) {
      this.inputRef.focus()
    }
  }

  setInputRef(element) {
    this.inputRef = element

    const { inputRef } = this.props
    if (typeof inputRef === 'function') {
      inputRef(element)
    }
  }

  render() {
    const {
      disabled,
      hasError,
      forceEncryptedPlaceholder,
      label,
      name,
      required,
      t,
      type
    } = this.props

    // Allow manifest to specify predefined label
    const localeKey =
      predefinedLabels.includes(label) || legacyLabels.includes(label)
        ? label
        : name

    // Cozy-UI <Field /> props
    const fieldProps = {
      ...this.props,
      autoComplete: 'off',
      className: 'u-m-0', // 0 margin
      disabled: disabled,
      error: !disabled && hasError,
      fullwidth: true,
      inputRef: this.setInputRef,
      label: t(`fields.${localeKey}.label`, {
        _: t(`legacy.fields.${localeKey}.label`, { _: name })
      }),
      placeholder: getFieldPlaceholder(
        this.props,
        t(`fields.${name}.placeholder`, { _: '' }),
        { forceEncryptedPlaceholder }
      ),
      side: required ? null : t('accountForm.fields.optional'),
      size: 'medium'
    }

    const passwordLabels = {
      hideLabel: t('accountForm.password.hide'),
      showLabel: t('accountForm.password.show')
    }

    switch (type) {
      case 'date':
        return (
          <Field
            {...fieldProps}
            placeholder={t(`fields.${name}.placeholder`, {
              _: t(`default.dateFormat`).toLowerCase()
            })}
          />
        )
      case 'dropdown':
        return <Field {...sanitizeSelectProps(fieldProps)} />
      case 'password':
        return (
          /*
          Using the `new-password` value is the best way to avoid
          autocomplete for password.
          See https://stackoverflow.com/a/17721462/1135990
           */
          <Field
            {...fieldProps}
            autoComplete="new-password"
            secondaryLabels={passwordLabels}
          />
        )
      default:
        return <Field {...fieldProps} type="text" />
    }
  }
}

AccountField.propTypes = {
  /**
   * The element wrapping the <Field /> component.
   * Passed to <SelectBox /> component.
   */
  container: PropTypes.instanceOf(Element),
  /**
   * Indicates if the <Field /> should be rendered with an error style.
   */
  hasError: PropTypes.bool,
  /**
   * Initial value of the field
   */
  initialValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * A callback to call on Field input ref.
   */
  inputRef: PropTypes.func,
  /**
   * Optionnal predefined label, used as locale key.
   */
  label: PropTypes.oneOf([...predefinedLabels, ...legacyLabels]),
  /**
   * field name as declared in manifest, used as locale key and as fallback for
   * label
   */
  name: PropTypes.string.isRequired,
  /**
   * Indicates if the field is required. Non-required fields will be indicated
   * as "optional"
   */
  required: PropTypes.bool,
  /**
   * Indicates fields role. Defined in manifest or during fields sanitization.
   * Valid values are ['identifier', 'password'].
   * The field having role identifier will be given focus.
   */
  role: PropTypes.string,
  /**
   * Field type, passed to <Field /> component. Except for "dropdown" which
   * will be mapped to "select"
   */
  type: PropTypes.oneOf(['date', 'dropdown', 'email', 'password', 'text']),
  /**
   * Translation function
   */
  t: PropTypes.func
}

export default translate()(AccountField)
