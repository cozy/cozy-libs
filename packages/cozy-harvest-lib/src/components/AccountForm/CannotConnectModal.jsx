import React from 'react'
import PropTypes from 'prop-types'

import Typography from 'cozy-ui/transpiled/react/Typography'
import { ConfirmDialog } from 'cozy-ui/transpiled/react/CozyDialogs'
import { useI18n } from 'cozy-ui/transpiled/react/I18n'

import { useComponentsProps } from '../Providers/ComponentsPropsProvider'
import Markdown from '../Markdown'

const CannotConnectModal = ({ vendorName, vendorLink, onClose }) => {
  const { t } = useI18n()
  const { ComponentsProps } = useComponentsProps()

  return (
    <ConfirmDialog
      open
      content={
        <>
          <Typography variant="h4" className="u-mb-1">
            {t('accountForm.cannotConnectModal.title')}
          </Typography>
          <Markdown
            className="u-mb-1"
            source={t('accountForm.cannotConnectModal.content', {
              vendorLink,
              vendorName
            })}
          />
          <Markdown
            className="u-mb-1"
            source={t('accountForm.cannotConnectModal.noAccount', {
              vendorLink,
              vendorName
            })}
          />
          {ComponentsProps?.CannotConnectModal?.extraContent}
        </>
      }
      onClose={onClose}
    />
  )
}

CannotConnectModal.propTypes = {
  vendorName: PropTypes.string,
  vendorLink: PropTypes.string,
  onClose: PropTypes.func
}

export default CannotConnectModal
