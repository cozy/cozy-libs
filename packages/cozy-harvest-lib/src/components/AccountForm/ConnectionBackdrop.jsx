import React, { useEffect } from 'react'
import { createPortal } from 'react-dom'

import { useWebviewIntent } from 'cozy-intent'
import Backdrop from 'cozy-ui/transpiled/react/Backdrop'
import LinearProgress from 'cozy-ui/transpiled/react/LinearProgress'
import Typography from 'cozy-ui/transpiled/react/Typography'
import useBreakpoints from 'cozy-ui/transpiled/react/providers/Breakpoints'
import CozyTheme from 'cozy-ui/transpiled/react/providers/CozyTheme'
import { useI18n } from 'cozy-ui/transpiled/react/providers/I18n'
import { makeStyles } from 'cozy-ui/transpiled/react/styles'

const useStyles = makeStyles({
  container: {
    position: 'fixed',
    zIndex: 'var(--zIndex-modal)',
    inset: 0
  }
})

export const ConnectionBackdrop = ({ name }) => {
  const styles = useStyles()
  const { isMobile } = useBreakpoints()
  const { t } = useI18n()

  return (
    <CozyTheme variant="inverted">
      <div className={styles.container}>
        <Backdrop
          open
          className="u-p-2"
          style={{
            backdropFilter: 'blur(4px)',
            WebkitBackdropFilter: 'blur(4px)' // Manually adding the -webkit- prefix as there's no CSS preprocessor
          }}
        >
          <div className="u-w-6 u-w-100-s">
            <Typography
              variant={isMobile ? 'h3' : 'h2'}
              className="u-ta-center"
            >
              {t('connectionBackdrop.connecting')}
            </Typography>

            <LinearProgress className="u-mt-1-half u-w-100" />

            <Typography
              variant={isMobile ? 'body2' : 'body1'}
              className="u-mt-1 u-ta-center"
            >
              {t('connectionBackdrop.progress', { name })}
            </Typography>
          </div>
        </Backdrop>
      </div>
    </CozyTheme>
  )
}

const ConnectionBackdropWrapper = ({ name }) => {
  const webviewIntent = useWebviewIntent()

  useEffect(() => {
    if (webviewIntent) {
      // On mounting, we want to ensure the UI displays white icone because the backdrop is dark
      // The WorkerView is then responsible to change back the icons to dark on white theme
      webviewIntent.call(
        'setFlagshipUI',
        {
          bottomTheme: 'light',
          topTheme: 'light'
        },
        'ConnectionBackdrop Start'
      )
    }

    // As a precaution, we set the icons back to dark on white theme when unmounting
    // This is to ensure the icons are not stuck in white on white theme in unhappy paths
    return () => {
      if (webviewIntent) {
        webviewIntent.call(
          'setFlagshipUI',
          {
            bottomTheme: 'dark',
            topTheme: 'dark'
          },
          'ConnectionBackdrop End'
        )
      }
    }
  }, [webviewIntent])

  return createPortal(<ConnectionBackdrop name={name} />, document.body)
}

export default ConnectionBackdropWrapper
