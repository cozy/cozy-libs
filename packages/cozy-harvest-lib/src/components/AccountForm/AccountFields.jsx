import PropTypes from 'react-proptypes'
import React, { PureComponent } from 'react'
import { Field as FinalFormField } from 'react-final-form'

import AccountField from './AccountField'
import { getEncryptedFieldName } from '../../helpers/fields'

// As SelectBox component from Cozy-UI, rendering dropdown type, is just giving
// us the full Option object, we just get its value to facilitate mapping
// with account
const parse = type => value => {
  return type === 'dropdown' ? value.value : value
}

export class AccountFields extends PureComponent {
  render() {
    const {
      container,
      disabled,
      hasError,
      initialValues,
      manifestFields,
      onKeyUp
    } = this.props

    // Ready to use named fields array
    const namedFields = Object.keys(manifestFields).map(fieldName => ({
      ...manifestFields[fieldName],
      name: fieldName
    }))

    return (
      <div onKeyUp={onKeyUp}>
        {namedFields.map((field, index) => (
          <FinalFormField
            key={index}
            name={field.name}
            parse={parse(field.type)}
          >
            {({ input }) => (
              <AccountField
                {...field}
                {...input}
                container={container}
                hasError={hasError}
                disabled={disabled}
                initialValue={
                  initialValues[field.name] ||
                  initialValues[getEncryptedFieldName(field.name)]
                }
              />
            )}
          </FinalFormField>
        ))}
      </div>
    )
  }
}

AccountFields.propTypes = {
  container: PropTypes.node,
  disabled: PropTypes.bool,
  hasError: PropTypes.bool,
  manifestFields: PropTypes.object.isRequired,
  onKeyUp: PropTypes.func
}

export default AccountFields
