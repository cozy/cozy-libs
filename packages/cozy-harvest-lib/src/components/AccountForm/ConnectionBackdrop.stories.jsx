import { ConnectionBackdrop } from './ConnectionBackdrop'

const meta = {
  component: ConnectionBackdrop,
  arg: {
    name: 'EDF'
  }
}

export default meta

export const Default = {
  name: 'Default'
}
