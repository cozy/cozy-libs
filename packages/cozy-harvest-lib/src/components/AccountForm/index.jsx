import React, { PureComponent } from 'react'
import { Form } from 'react-final-form'
import PropTypes from 'prop-types'

import { isMobile } from 'cozy-device-helper'
import Button from 'cozy-ui/transpiled/react/Button'
import { translate } from 'cozy-ui/transpiled/react/I18n'

import AccountFields from './AccountFields'
import TriggerErrorInfo from '../infos/TriggerErrorInfo'
import { getEncryptedFieldName } from '../../helpers/fields'
import { KonnectorJobError } from '../../helpers/konnectors'
import manifest from '../../helpers/manifest'
import withKonnectorLocales from '../hoc/withKonnectorLocales'

const VALIDATION_ERROR_REQUIRED_FIELD = 'VALIDATION_ERROR_REQUIRED_FIELD'

/**
 * AccountForm is reponsible of generating a form which will allow user to
 * edit an account for a given konnector.
 *
 * This component rely internally on a ReactFinalForm component.
 *
 * AccountForm must be passed `onSubmit` callback.
 * @type {PureComponent}
 * @see https://github.com/final-form/react-final-form#getting-started
 */
export class AccountForm extends PureComponent {
  constructor(props) {
    super(props)

    this.inputs = {}
    this.inputFocused = null

    this.inputRefByName = this.inputRefByName.bind(this)
    this.handleFocus = this.handleFocus.bind(this)
    this.handleKeyUp = this.handleKeyUp.bind(this)
    this.focusNext = this.focusNext.bind(this)
  }

  /**
   * Indicates if the state of ReactFinalForm implies that data can be submitted
   * @param  {Object} formState See https://github.com/final-form/final-form#formstate
   * @return {Boolean}
   */
  isSubmittable({ dirty, error, initialValues, valid }) {
    const modified = !initialValues || dirty
    // Here error is not a validation error but an instance of a
    // KonnectorJobError, so submitting again is possible
    return modified && (error || valid)
  }

  /**
   * Give focus to next input in the form.
   * Fallback for mobile devices as we are using a div element instead of a
   * form.
   * @return {Element} Focused element or null if no element has been focused
   */
  focusNext() {
    if (!this.inputs) return null

    const inputs = Object.values(this.inputs)

    const currentIndex = inputs.indexOf(this.inputFocused)

    let nextIndex = currentIndex + 1
    let nextInput = inputs[nextIndex]

    if (nextInput) {
      nextInput.focus()
    }

    return nextInput || null
  }

  /**
   * Capture input with focus and store the current focused input
   * @param  {Event} event Focus event
   */
  handleFocus(event) {
    if (Object.values(this.inputs).includes(event.target)) {
      this.inputFocused = event.target
    }
  }

  /**
   * Handle key up and check if `ENTER` key has been pressed. If so, submit the
   * form if all parmater are ok.
   * @param  {React.SyntheticEvent} event         Key events
   * @param  {Boolean} dirty         Indicates if form is dirty, i.e. if values
   * have changed
   * @param  {Object} form          The form object injected by ReactFinalForm.
   * @param  {Object} initialValues Initial data injected into AccountForm
   * @param  {Boolean} valid        Indicates if the form data is valid
   * @param  {Object} values        Actual form values data
   */
  handleKeyUp(event, { dirty, form, initialValues, valid, values }) {
    if (event.key === 'Enter') {
      const changedFocus = isMobile() && !!this.focusNext()
      if (
        !changedFocus &&
        this.isSubmittable({ dirty, initialValues, valid })
      ) {
        this.handleSubmit(values, form)
      }
    }
  }

  /**
   * Submit handler
   * @param  {Object} values        Actual form values data
   * @param  {Object} form          The form object injected by ReactFinalForm.
   */
  handleSubmit(values, form) {
    const { onSubmit } = this.props
    // Reset form with new values to set back dirty to false
    form.reset(values)
    onSubmit(values)
  }

  /**
   * Callback passed to `<AccountFields />` element. Called with a field name,
   * it return a callback wich take an input ref as argument. It then stores
   * the input ref into AccountForm's inner `inputs` property.
   * With this method we keep a list of inputs existing in the form.
   * The inputs inner property is indexed with names to avoid duplicates in case
   * of re-render of input children.
   * @param  {string} fieldName Field name
   * @return {function}         Callback which take an input ref as argument.
   */
  inputRefByName(fieldName) {
    return input => {
      this.inputs[fieldName] = input
    }
  }

  validate = (fields, initialValues) => vals => {
    let errors = {}
    for (let name in fields)
      if (
        fields[name].required &&
        !vals[name] &&
        // Don't require value for empty encrypted fields with initial value
        !initialValues[getEncryptedFieldName(name)]
      )
        errors[name] = VALIDATION_ERROR_REQUIRED_FIELD
    return errors
  }

  render() {
    const {
      account,
      error,
      konnector,
      onSubmit,
      showError,
      submitting,
      t
    } = this.props
    const { fields } = konnector

    const sanitizedFields = manifest.sanitizeFields(fields)
    const defaultValues = manifest.defaultFieldsValues(sanitizedFields)
    const initialValues = account && account.auth
    const initialAndDefaultValues = { ...defaultValues, ...initialValues }

    let container = null
    console.log('error AccountFrom', error)
    const isLoginError =
      error instanceof KonnectorJobError && error.isLoginError()

    return (
      // See https://github.com/final-form/react-final-form#getting-started
      <Form
        initialValues={initialAndDefaultValues}
        onSubmit={onSubmit}
        validate={this.validate(sanitizedFields, initialAndDefaultValues)}
        render={({ dirty, form, values, valid }) => (
          <div
            onKeyUp={event =>
              this.handleKeyUp(event, {
                dirty,
                form,
                initialValues,
                valid,
                values
              })
            }
            onFocusCapture={this.handleFocus}
            ref={element => {
              container = element
            }}
          >
            {error && showError && (
              <TriggerErrorInfo
                className="u-mb-1"
                error={error}
                konnector={konnector}
              />
            )}
            <AccountFields
              container={container}
              disabled={submitting}
              fields={sanitizedFields}
              hasError={error && isLoginError}
              initialValues={initialAndDefaultValues}
              inputRefByName={this.inputRefByName}
              t={t}
            />
            <Button
              busy={submitting}
              className="u-mt-2 u-mb-1-half"
              disabled={
                submitting ||
                !this.isSubmittable({ dirty, error, initialValues, valid })
              }
              extension="full"
              label={t('accountForm.submit.label')}
              onClick={() => this.handleSubmit(values, form)}
            />
          </div>
        )}
      />
    )
  }
}

AccountForm.propTypes = {
  /**
   * Account document. Used to get intial form values.
   * If no account is passed, AccountForm will use empty initial values.
   * @type {Object}
   */
  account: PropTypes.object,
  /**
   * Existing error
   * @type {Error,KonnectorJobError}
   */
  error: PropTypes.object,
  /**
   * Konnector document. AccountForm will check the `fields` object to compute
   * fields.
   * @type {Object}
   */
  konnector: PropTypes.object.isRequired,
  /**
   * Submit callback
   * @type {Function}
   */
  onSubmit: PropTypes.func.isRequired,
  /**
   * Indicates if the AccountForm has to show errors. Sometimes errors may be
   * displayed elsewhere. However, a KonnectorJobError corresponding to a login
   * error is always displayed.
   * @type {Boolean}
   */
  showError: PropTypes.bool,
  /**
   * Indicates if the form should be rendered as submitting data or busy.
   * Typically updated after an `onSubmit` call.
   * @type {Object}
   */
  submitting: PropTypes.bool,
  /**
   * Translation function
   */
  t: PropTypes.func
}

AccountForm.defaultProps = {
  showError: true
}

export default translate()(withKonnectorLocales(AccountForm))
