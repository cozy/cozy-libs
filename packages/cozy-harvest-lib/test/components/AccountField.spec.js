import React from 'react'
import { shallow } from 'enzyme'
import PropTypes from 'prop-types'

import { AccountField } from 'components/AccountForm/AccountField'

const fixtures = {
  username: {
    encrypted: false,
    required: true,
    type: 'text'
  },
  passphrase: {
    encrypted: true,
    required: true,
    type: 'password'
  }
}

const t = jest.fn()

describe('AccountField', () => {
  beforeEach(() => {
    t.mockClear()
    t.mockImplementation(key => key)
  })

  it('should render', () => {
    const wrapper = shallow(
      <AccountField {...fixtures.username} name="username" t={t} />
    )
    const component = wrapper.dive().getElement()
    expect(component).toMatchSnapshot()
  })

  it('render a date field', () => {
    const wrapper = shallow(
      <AccountField required={true} type="date" name="date" t={t} />
    )
    const component = wrapper.dive().getElement()
    expect(component).toMatchSnapshot()
  })

  it('render a password field', () => {
    const wrapper = shallow(
      <AccountField {...fixtures.passphrase} name="passphrase" t={t} />
    )
    const component = wrapper.dive().getElement()
    expect(component).toMatchSnapshot()
  })

  it('hide the password toggle on empty fields', () => {
    const wrapper = shallow(
      <AccountField {...fixtures.passphrase} name="passphrase" t={t} value="" />
    )
    const component = wrapper.dive().dive()
    expect(component.prop('side')).toBe(null)
  })

  it('show the password toggle on non-empty fields', () => {
    const wrapper = shallow(
      <AccountField
        {...fixtures.passphrase}
        name="passphrase"
        value="123"
        t={t}
      />
    )
    const component = wrapper.dive().dive()
    expect(component.prop('side')).not.toBe(null)
  })

  it('render a dropdown field', () => {
    const options = [
      { label: 'Option 1', value: 'option1' },
      { label: 'Option 2', value: 'option2' }
    ]

    const wrapper = shallow(
      <AccountField
        name="multiple"
        options={options}
        required={true}
        t={t}
        type="dropdown"
      />
    )
    const component = wrapper.dive().getElement()
    expect(component).toMatchSnapshot()
  })

  it('uses predefined label', () => {
    const wrapper = shallow(
      <AccountField
        label="login"
        name="username"
        required={true}
        type="text"
        t={t}
      />
    )
    expect(wrapper.props().label).toBe('fields.login.label')
  })

  it('ignores invalid predefined label', () => {
    // Avoid warning
    class AccountFieldWithPermissiveLabel extends AccountField {}
    AccountFieldWithPermissiveLabel.propTypes = {
      ...AccountField.propTypes,
      label: PropTypes.string
    }

    const wrapper = shallow(
      <AccountFieldWithPermissiveLabel
        label="foo"
        name="username"
        required={true}
        type="text"
        t={t}
      />
    )
    expect(wrapper.props().label).toBe('fields.username.label')
  })
})
