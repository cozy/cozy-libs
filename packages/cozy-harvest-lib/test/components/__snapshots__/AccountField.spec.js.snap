// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountField render a date field 1`] = `
<div
  className="styles__o-field___3n5HM u-m-0"
>
  <Label
    block={true}
    className=""
    error={false}
    htmlFor="harvest-account-date"
  >
    fields.date.label
  </Label>
  <Input
    autoComplete="off"
    error={false}
    fullwidth={true}
    id="harvest-account-date"
    inputRef={[Function]}
    name="date"
    placeholder="fields.date.placeholder"
    size="medium"
    type="date"
  />
</div>
`;

exports[`AccountField render a dropdown field 1`] = `
<div
  className="styles__o-field___3n5HM u-m-0"
>
  <Label
    block={true}
    className=""
    error={false}
    htmlFor="harvest-account-multiple"
  >
    fields.multiple.label
  </Label>
  <withBreakpoints(SelectBox)
    className="u-m-0"
    error={false}
    fieldProps={Object {}}
    fullwidth={true}
    id="harvest-account-multiple"
    inputRef={[Function]}
    label="fields.multiple.label"
    labelProps={Object {}}
    name="multiple"
    options={
      Array [
        Object {
          "label": "Option 1",
          "value": "option1",
        },
        Object {
          "label": "Option 2",
          "value": "option2",
        },
      ]
    }
    placeholder="fields.multiple.placeholder"
    required={true}
    secondaryLabels={Object {}}
    side={null}
    size="medium"
    t={
      [MockFunction] {
        "calls": Array [
          Array [
            "legacy.fields.multiple.label",
            Object {
              "_": "multiple",
            },
          ],
          Array [
            "fields.multiple.label",
            Object {
              "_": "legacy.fields.multiple.label",
            },
          ],
          Array [
            "fields.multiple.placeholder",
            Object {
              "_": "",
            },
          ],
          Array [
            "accountForm.password.hide",
          ],
          Array [
            "accountForm.password.show",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": "legacy.fields.multiple.label",
          },
          Object {
            "type": "return",
            "value": "fields.multiple.label",
          },
          Object {
            "type": "return",
            "value": "fields.multiple.placeholder",
          },
          Object {
            "type": "return",
            "value": "accountForm.password.hide",
          },
          Object {
            "type": "return",
            "value": "accountForm.password.show",
          },
        ],
      }
    }
    type="select"
  />
</div>
`;

exports[`AccountField render a password field 1`] = `
<div
  className="styles__o-field___3n5HM u-m-0"
>
  <Label
    block={true}
    className=""
    error={false}
    htmlFor="harvest-account-passphrase"
  >
    fields.passphrase.label
  </Label>
  <InputPassword
    autoComplete="new-password"
    error={false}
    fullwidth={true}
    hideLabel="accountForm.password.hide"
    id="harvest-account-passphrase"
    inputRef={[Function]}
    name="passphrase"
    placeholder="fields.passphrase.placeholder"
    showLabel="accountForm.password.show"
    size="medium"
    type="password"
  />
</div>
`;

exports[`AccountField should render 1`] = `
<div
  className="styles__o-field___3n5HM u-m-0"
>
  <Label
    block={true}
    className=""
    error={false}
    htmlFor="harvest-account-username"
  >
    fields.username.label
  </Label>
  <Input
    autoComplete="off"
    error={false}
    fullwidth={true}
    id="harvest-account-username"
    inputRef={[Function]}
    name="username"
    placeholder="fields.username.placeholder"
    size="medium"
    type="text"
  />
</div>
`;
